Web Manifest: support id member

marcoscaceres opened the pull request:
#### 7e448378499b72aa612f9dd7fdeb1dc7f7cd1b25
<pre>
Web Manifest: support id member
<a href="https://bugs.webkit.org/show_bug.cgi?id=230596">https://bugs.webkit.org/show_bug.cgi?id=230596</a>
rdar://83656112

Implementation of Web Manifest&apos;s id member:
<a href="https://www.w3.org/TR/appmanifest/#id-member">https://www.w3.org/TR/appmanifest/#id-member</a>

The manifest&apos;s id member is a string that represents the identity for the application. The identity takes the form of a URL, which is same origin as the start URL.

Reviewed by Devin Rousso.

* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseId):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest manifestId]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(ApplicationManifestParserTest::testId):
(assertManifestHasDefaultValues):
(TEST_F):

Canonical link: <a href="https://commits.webkit.org/256182@main">https://commits.webkit.org/256182@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/165b16fb63c290efad7934e1fde57eb7ca0743be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94965 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4111 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27871 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104568 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164831 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98959 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4196 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32294 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87269 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100482 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100633 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3036 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81446 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30016 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84951 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72905 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38703 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18338 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36531 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19619 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4265 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40459 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42436 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38853 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -80,6 +80,8 @@ ApplicationManifest ApplicationManifestParser::parseManifest(const String& text,
     parsedManifest.themeColor = parseColor(*manifest, "theme_color"_s);
     parsedManifest.icons = parseIcons(*manifest);
+    // this is "id" in the spec, but we use "appId" because "id" mean something special in ObjC land.
+    parsedManifest.appId = parseId(*manifest, parsedManifest.startURL);
```
i think we should have this as `id` for everything in C++.  it's only `_WKApplicationManifest` that might need to worry about this (and even then maybe something like `idValue` or `manifestId` would be more of an apt name 🤔)

dcrousso commented on Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
```diff
@@ -58,6 +58,7 @@ WK_CLASS_AVAILABLE(macos(10.13.4), ios(11.3))
 @property (nonatomic, readonly, copy) NSURL *startURL;
 @property (nonatomic, readonly) _WKApplicationManifestDisplayMode displayMode;
 @property (nonatomic, readonly, copy) NSArray<_WKApplicationManifestIcon *> *icons WK_API_AVAILABLE(macos(13.0), ios(16.0));
+@property (nonatomic, readonly, nullable, copy) NSURL *appId;
```
this needs a `WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))`

dcrousso commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
```diff
@@ -63,6 +63,7 @@ struct ApplicationManifest {
     URL startURL;
     Color themeColor;
     Vector<Icon> icons;
+    URL appId;
```
NIT: I'd maybe move this to be closer to the `startURL` since the two are kinda intertwined

ditto below with all the other spots too

dcrousso commented on Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
```diff
@@ -56,6 +56,7 @@ WK_CLASS_AVAILABLE(macos(10.13.4), ios(11.3))
 @property (nonatomic, readonly, nullable, copy) NSString *applicationDescription;
 @property (nonatomic, readonly, nullable, copy) NSURL *scope;
 @property (nonatomic, readonly, copy) NSURL *startURL;
+@property (nonatomic, readonly, copy) NSURL *manifestId WK_API_AVAILABLE(macos(13.0), ios(16.0));
```
this should be `WK_MAC_TBA` and `WK_IOS_TBA` instead of the `13.0` and `16.0` respectively

marcoscaceres replied:
fixed! sorry I missed that in the original review.

marcoscaceres committed:
Web Manifest: support id member
https://bugs.webkit.org/show_bug.cgi?id=230596
rdar://83656112

Implementation of Web Manifest's id member:
https://www.w3.org/TR/appmanifest/#id-member

The manifest's id member is a string that represents the identity for the application. The identity takes the form of a URL, which is same origin as the start URL.

Reviewed by Devin Rousso.

* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseId):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest manifestId]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(ApplicationManifestParserTest::testId):
(assertManifestHasDefaultValues):
(TEST_F):

Canonical link: https://commits.webkit.org/256182@main
